### PR TITLE
Add Regal language server support

### PR DESCRIPTION
**Pull request by charlieegan3**
This is WIP but getting there. There are a few changes here:

* The extension has been extended to support one or more language servers for Rego. Currently only [Regal](https://github.com/StyraInc/regal). These are started based on configuration and can also be installed if the user desires.
* The OPA installation logic and downloader has been made generic to support fetching of regal and other binaries as needed in future.
* Language servers are promoted to users if they are not installed.

Some suggested follow on work:

* make binary dependency checking generic, and use server module for OPA checking
* make finding of binary paths more generic

TODOs still include:

- [x] testing of this change on a windows machine
- [ ] testing of this change on a linux machine
- [x] Updating Regal and sharing a patch release with a minor LS change (0.16.1) (https://github.com/StyraInc/regal/pull/539) and https://github.com/StyraInc/regal/pull/569

Some videos of common flows:

Regal needs update:

https://github.com/open-policy-agent/vscode-opa/assets/1774239/d0559cc9-5eab-4c14-97e5-6fac6b49d8f2

Regal install needed:

https://github.com/open-policy-agent/vscode-opa/assets/1774239/f9b76d8a-11dc-4e0d-b50b-97d5a4fe892c

Regal installed already:

https://github.com/open-policy-agent/vscode-opa/assets/1774239/75c375ad-42c9-4159-bbc9-f6852a88b085

### CODE REVIEW

File: src/github-installer.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                    determineBinaryURLFromRelease,
+                    determineExecutableName,
+                );
+            } else {
+                installDeclined[binaryConfigKey] = true;
+            }
+        });
+}
+
+async function install(binaryConfigKey: string, repo: string, determineBinaryURLFromRelease: (release: any) => string, determineExecutableName: () => string) {
```
Comment (anderseknert) [inline src/github-installer.ts:41]: Very imperative, but also very easy to follow!

File: src/github-installer.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            });
+        }).on('error', (e: any) => reject(e));
+    });
+}
+
+function getToken(): string {
+    // Need an OAuth token to access Github because
+    // this gets around the ridiculously low
+    // anonymous access rate limits (60 requests/sec/IP)
+    // This token only gives access to "public_repo" and "repo:status" scopes
```
Comment (anderseknert) [inline src/github-installer.ts:136]: This comment certainly is needed! 😄

**Review by anderseknert — APPROVED** [review]
Awesome! Really easy to follow along in this code. I just had a question or two.

File: src/github-installer.ts (hunk 2 above)
Comment (charlieegan3) [reply src/github-installer.ts:136]: Ha indeed, it's copied from the old implementation however and so isn't new to this PR.

File: src/github-installer.ts (hunk 1 above)
Comment (charlieegan3) [reply src/github-installer.ts:41]: it does the job, mostly just copy pasted from the old OPA installer, with some parts made more generic.

File: package.json [hunk 3]
```diff
@@ -10,14 +10,11 @@
...
 	"publisher": "tsandall",
 	"engines": {
-		"vscode": "^1.21.0"
+		"vscode": "^1.75.0"
 	},
 	"categories": [
 		"Programming Languages"
 	],
-	"activationEvents": [
-		"onLanguage:rego"
```
[the remark this answers is not in the document]
Comment (anderseknert) [reply package.json:19]: Are we sure all users are on modern VS Code versions then? 🙂 I'd just leave it in unless it's ancient.

[the remark this answers is not in the document]
Comment (charlieegan3) [reply package.json:19]: > This activation event can be removed as VS Code generates these automatically from your package.json contribution declarations.

This is the full message. My understanding is that this applies from 1.75.0, which is over a year old. https://github.com/microsoft/vscode/pull/173345

[the remark this answers is not in the document]
Comment (anderseknert) [reply package.json:19]: Oh well, I guess we should listen to the linters 🙂

**Comment by charlieegan3** [comment]
https://github.com/StyraInc/regal/pull/569 is merged now so all that's needed is a Regal release and to update https://github.com/open-policy-agent/vscode-opa/pull/98/files#diff-1944111a462efb504e2d39ee24a9cb532c03164ada37caada104fecc06d874bdR23.

**Pull request merged**